flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/rank/AppendRankFunction.java

```diff
@@ -0,0 +1,228 @@
+	private transient TopNBuffer buffer;
+
+	// the kvSortedMap stores mapping from partition key to it's buffer
+	private transient Map<BaseRow, TopNBuffer> kvSortedMap;
+
+	public AppendRankFunction(long minRetentionTime, long maxRetentionTime, BaseRowTypeInfo inputRowType,
+			GeneratedRecordComparator sortKeyGeneratedRecordComparator, BaseRowKeySelector sortKeySelector,
+			RankType rankType, RankRange rankRange, GeneratedRecordEqualiser generatedEqualiser,
+			boolean generateRetraction, boolean outputRankNumber, long cacheSize) {
+		super(minRetentionTime, maxRetentionTime, inputRowType, sortKeyGeneratedRecordComparator, sortKeySelector,
+				rankType, rankRange, generatedEqualiser, generateRetraction, outputRankNumber);
+		this.sortKeyType = sortKeySelector.getProducedType();
+		this.inputRowSer = inputRowType.createSerializer(new ExecutionConfig());
+		this.cacheSize = cacheSize;
+	}
+
+	public void open(Configuration parameters) throws Exception {
+		super.open(parameters);
+		int lruCacheSize = Math.max(1, (int) (cacheSize / getDefaultTopNSize()));
+		kvSortedMap = new LRUMap<>(lruCacheSize);
+		LOG.info("Top{} operator is using LRU caches key-size: {}", getDefaultTopNSize(), lruCacheSize);
+
+		ListTypeInfo<BaseRow> valueTypeInfo = new ListTypeInfo<>(inputRowType);
+		MapStateDescriptor<BaseRow, List<BaseRow>> mapStateDescriptor = new MapStateDescriptor(
+				"data-state-with-append", sortKeyType, valueTypeInfo);
+		dataState = getRuntimeContext().getMapState(mapStateDescriptor);
+
+		// metrics
+		registerMetric(kvSortedMap.size() * getDefaultTopNSize());
+	}
+
+	@Override
+	public void processElement(BaseRow input, Context context, Collector<BaseRow> out) throws Exception {
+		long currentTime = context.timerService().currentProcessingTime();
+		// register state-cleanup timer
+		registerProcessingCleanupTimer(context, currentTime);
+
+		initHeapStates();
+		initRankEnd(input);
+
+		BaseRow sortKey = sortKeySelector.getKey(input);
+		// check whether the sortKey is in the topN range
+		if (checkSortKeyInBufferRange(sortKey, buffer)) {
+			// insert sort key into buffer
+			buffer.put(sortKey, inputRowSer.copy(input));
+			Collection<BaseRow> inputs = buffer.get(sortKey);
+			// update data state
+			dataState.put(sortKey, (List<BaseRow>) inputs);
+			if (outputRankNumber || hasOffset()) {
+				// the without-number-algorithm can't handle topN with offset,
+				// so use the with-number-algorithm to handle offset
+				processElementWithRowNumber(sortKey, input, out);
+			} else {
+				processElementWithoutRowNumber(input, out);
+			}
+		}
+	}
+
+	@Override
+	public void onTimer(
+			long timestamp,
+			OnTimerContext ctx,
+			Collector<BaseRow> out) throws Exception {
+		if (stateCleaningEnabled) {
+			// cleanup cache
+			kvSortedMap.remove(keyContext.getCurrentKey());
+			cleanupState(dataState);
+		}
+	}
+
+	@Override
+	protected long getMaxSizeOfBuffer() {
```

Comment:
This method is introduced in `ApproximateRank` and I don't think we need this anymore. We can use `getDefaultTopNSize()` instead of this method is invoked.